### PR TITLE
[Woo] Support two networking implementations for REST API project

### PR DESCRIPTION
**Pull request by hichamboushaba**
⚠️ Please don't merge until #2594 is merged.

To test, repeat the test steps of #2587.

### CODE REVIEW

File: config/checkstyle.xml [hunk 1]
```diff
@@ -95,7 +95,7 @@
 
     <!-- Specific to FluxC - Kotlin -->
     <module name="RegexpMultiline">
-        <property name="format" value="(?&lt;!@Singleton)\n.*class(.+\n)*.*:[\n\s]*Base[\w]*Client\("/>
+        <property name="format" value="(?&lt;!@Singleton)\n.*^class(.+\n)*.*:[\n\s]*Base[\w]*Client\("/>
```
Comment (hichamboushaba) [inline config/checkstyle.xml:98]: Check the commit message for why I made this change.

**Review by hafizrahman — COMMENTED** [review]
Nice work! My general sense (just from a workload perspective) with this implementation is that it won't require a lot more time compared if we're going with single networking implementation.

Do we also want to handle cases where at some point users might turn on/off Jetpack or turn on/off Application Passwords, when they're already using the app?

**Comment by hichamboushaba** [comment]
Thanks for the review @hafizrahman 

> Do we also want to handle cases where at some point users might turn on/off Jetpack or turn on/off Application Passwords, when they're already using the app?

Since the decision on what path to take depends only on the instance of the `SiteModel` we pass, we can support any scenario we want in the app by passing the correct one:
1. If the site gets connected to Jetpack, we will just need to fetch sites using the traditional way `/me/sites`, then update the `SelectedSite` using the new instance, which should work automatically.
2. If there are any plans to support the other way around, we will do the same thing after requesting the user's wp-admin credentials.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/WooNetwork.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    suspend fun <T : Any> executeGetGsonRequest(
+        site: SiteModel,
+        path: String,
+        clazz: Class<T>,
+        params: Map<String, String> = emptyMap()
+    ): WPAPIResponse<T> {
+        return when (site.origin) {
+            SiteModel.ORIGIN_WPCOM_REST -> jetpackTunnelWPAPINetwork.executeGetGsonRequest(site, path, clazz, params)
+                .toWPAPIResponse()
+            else -> applicationPasswordNetwork.executeGetGsonRequest(site, path, clazz, params)
```
Comment (hafizrahman) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/WooNetwork.kt]: probably np: should this handle only `SiteModel.ORIGIN_XMLRPC`, and add error handling if it's `SiteModel.ORIGIN_UNKNOWN`?

As far as I can check, the value is always eventually set to either `ORIGIN_WPCOM_REST` or `ORIGIN_XMLRPC` in WooCommerce app's FluxC usage, so this should be OK. Is it worth being extra safe here?

Comment (hichamboushaba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/WooNetwork.kt]: Good point, done in d209e60

**Review by hafizrahman — APPROVED** [review]

**Pull request merged**